### PR TITLE
pyocf: volume changes required for failover tests

### PR DESCRIPTION
**Pull request by arutk**

### CODE REVIEW

**Comment by arutk** [comment]
retest this please

**Review by mmichal10 — DISMISSED** [review]

**Review by robertbaldyga — CHANGES_REQUESTED** [review]

File: tests/functional/pyocf/types/volume_replicated.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+        @CFUNCTYPE(None, c_void_p, c_int)
+        def cb(io, err):
+            nonlocal io_remaining
+            nonlocal error
+            nonlocal original_cb
+            nonlocal lock
+            io = cast(io, POINTER(Io))
+
+            lock.acquire()
```
Comment (jfckm) [inline tests/functional/pyocf/types/volume_replicated.py:57]: I think it would be more readable as context manager:
```python
 with lock:
            if err:
                error = err
            io_remaining -= 1
            finished = True if io_remaining == 0 else False
```

**Review by jfckm — CHANGES_REQUESTED** [review]

File: tests/functional/pyocf/types/volume.py [hunk 2]
```diff
@@ -272,6 +274,34 @@ def dump(self, offset=0, size=0, ignore=VOLUME_POISON, **kwargs):
...
+        self.offlined = False
+
+    def _reject_io(self, io):
+        cast(io, POINTER(Io)).contents._end(io, -OcfErrorCode.OCF_ERR_IO)
+
+    def submit_flush_check_offline(self, io):
+        if self.offlined:
+            self._reject_io(io)
+        else:
+            self.submit_flush(io)
```
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/pyocf/types/volume.py:312]: what do you mean?

File: tests/functional/pyocf/types/cache.py [hunk 3]
```diff
@@ -659,6 +659,19 @@ def remove_core(self, core: Core):
...
+    def new_io(
+        self, queue: Queue, addr: int, length: int, direction: IoDir,
+        io_class: int, flags: int
+    ):
+        lib = OcfLib.getInstance()
+        volume = lib.ocf_cache_get_front_volume(self.cache_handle)
+        io = lib.ocf_volume_new_io(
+            volume, queue.handle, addr, length, direction, io_class, flags)
+        return Io.from_pointer(io)
+
```
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/pyocf/types/cache.py:674]: ok, removed new_io() from cache

File: tests/functional/pyocf/types/cache.py [hunk 4]
```diff
@@ -659,6 +659,19 @@ def remove_core(self, core: Core):
 
         self.cores.remove(core)
 
+    def get_volume(self):
```
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/pyocf/types/cache.py:662]: done

File: tests/functional/pyocf/types/io.py [hunk 5]
```diff
@@ -98,6 +98,33 @@ def end(self, err):
     def submit(self):
         return OcfLib.getInstance().ocf_core_submit_io_wrapper(byref(self))
 
+    def submit_sync(self):
+        finished = False
+        error = 0
+        cond = Condition()
+
+        @CFUNCTYPE(c_void_p, c_int)
+        def cb(err):
```
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/pyocf/types/io.py:107]: done

File: tests/functional/pyocf/types/io.py [hunk 6]
```diff
@@ -98,6 +98,33 @@ def end(self, err):
     def submit(self):
         return OcfLib.getInstance().ocf_core_submit_io_wrapper(byref(self))
 
+    def submit_sync(self):
+        finished = False
+        error = 0
+        cond = Condition()
```
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/pyocf/types/io.py:104]: removed this entirely, this is no loner needed in current version

File: tests/functional/pyocf/types/volume.py [hunk 7]
```diff
@@ -101,7 +99,7 @@ def __init__(self, size: S, uuid=None):
 
         self.data = create_string_buffer(int(self.size))
         memset(self.data, VOLUME_POISON, self.size)
-        self._storage = cast(self.data, c_void_p)
+        self.data_ptr = cast(self.data, c_void_p).value
```
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/pyocf/types/volume.py]: Don't know why, but this is needed. Volume::resize() is most likely broken. I will add ".value" there

File: tests/functional/tests/management/test_attach_cache.py [hunk 8]
```diff
@@ -32,6 +32,7 @@
 
 logger = logging.getLogger(__name__)
 
+import pdb
```
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/tests/management/test_attach_cache.py:35]: fixed

File: tests/functional/pyocf/types/volume_exp_obj.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+from pyocf.utils import Size
+from pyocf.types.data import Data
+from pyocf.types.io import IoDir, Io
+from pyocf.types.shared import OcfCompletion
+
+
+class ExpObjVolume(Volume):
+    def __init__(self, cc, uuid=None):
+        super().__init__(uuid)
+        self.cc = cc
```
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/pyocf/types/volume_exp_obj.py:19]: ok, changed to "parent"

File: tests/functional/pyocf/types/volume_exp_obj.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+            io.contents._addr,
+            io.contents._bytes,
+            io.contents._dir,
+            io.contents._class,
+            io.contents._flags,
+        )
+
+        lib = OcfLib.getInstance()
+        cdata = OcfLib.getInstance().ocf_io_get_data(io)
+        OcfLib.getInstance().ocf_io_set_data(byref(exp_obj_io), cdata, 0)
```
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/pyocf/types/volume_exp_obj.py:38]: OCF API does not allow to set offset != 0. Currently this is only used for OCF to seek for the purose of ctx_data_read or ctx_data_write

File: tests/functional/pyocf/types/volume_replicated.py (hunk 1 above)
Comment (arutk) [reply tests/functional/pyocf/types/volume_replicated.py:57]: done

File: tests/functional/pyocf/types/volume.py [hunk 11]
```diff
@@ -272,6 +274,34 @@ def dump(self, offset=0, size=0, ignore=VOLUME_POISON, **kwargs):
     def md5(self):
         pass
 
+    def offline(self):
+        self.offlined = True
```
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/pyocf/types/volume.py:300]: sure, done

File: tests/functional/pyocf/types/volume.py [hunk 12]
```diff
@@ -239,6 +256,7 @@ def __init__(self, uuid=None):
         type(self)._uuid_[self.uuid] = self
 
         self.offlined = False
+        self.opened = False
```
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/pyocf/types/volume.py:263]: fixed

File: tests/functional/pyocf/types/volume.py [hunk 13]
```diff
@@ -77,65 +77,45 @@ class VolumeIoPriv(Structure):
...
+            _volume_priv_size=0,
+            _caps=VolumeCaps(_atomic_writes=0),
+            _ops=cls.get_ops(),
+            _io_ops=cls.get_io_ops(),
+            _deinit=0,
+        )
         return cls.props
 
+    def get_copy(self):
+        pass
```
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/pyocf/types/volume.py:138]: done

File: tests/functional/pyocf/types/volume_cache.py [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    def submit_flush(self, flush):
+        io = cast(flush, POINTER(Io))
+        io.contents._end(io, 0)
+
+    def get_stats(self):
+        # TODO: how?
+        pass
+
+    def reset_stats(self):
+        pass
```
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/pyocf/types/volume_cache.py:30]: done

File: tests/functional/pyocf/types/volume.py [hunk 15]
```diff
@@ -100,35 +102,37 @@ def unregister_volume_type(cls, ctx):
...
-            _submit_discard=cls._submit_discard,
-            _submit_write_zeroes=cls._submit_write_zeroes,
-            _open=cls._open,
-            _close=cls._close,
-            _get_max_io_size=cls._get_max_io_size,
-            _get_length=cls._get_length,
-        )
-        return cls.ops
+        if cls not in Volume.ops.keys():
+            Volume.ops[cls] = VolumeOps(
```
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/pyocf/types/volume.py:106]: this commit is an artifact from earlier implementation

[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/pyocf/types/volume.py:106]: will be reworked in next update

File: tests/functional/pyocf/types/cache.py [hunk 16]
```diff
@@ -659,6 +659,19 @@ def remove_core(self, core: Core):
...
+        return lib.ocf_cache_get_front_volume(self.cache_handle)
+
+    def new_io(
+        self, queue: Queue, addr: int, length: int, direction: IoDir,
+        io_class: int, flags: int
+    ):
+        lib = OcfLib.getInstance()
```
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/pyocf/types/cache.py:669]: this code will be removed in next update (replaced with volume_new_io)

File: tests/functional/pyocf/types/volume.py [hunk 17]
```diff
@@ -81,6 +83,21 @@ class Volume:
...
+            ctx = OcfCtx.get_default()
+        props = byref(cls.get_props())
+        ctx.register_volume_type(Volume.type_id, props)
+
+    @classmethod
+    def unregister_volume_type(cls, ctx):
+        if not ctx:
+            ctx = OcfCtx.get_default()
+        ctx.unregister_volume_type(Volume.type_id)
+
```
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/pyocf/types/volume.py]: changed this in the latest version according to our discussion

File: tests/functional/pyocf/types/core.py [hunk 18]
```diff
@@ -114,6 +114,10 @@ def new_io(
 
         return Io.from_pointer(io)
 
+    def get_front_volume(self):
+        assert self.device == lib.ocf_core_get_volume(self.handle)
+        return lib.ocf_core_get_volume(self.handle)
```
Comment (robertbaldyga) [inline tests/functional/pyocf/types/core.py:119]: `ocf_core_get_front_volume()`?

File: tests/functional/pyocf/types/volume_cache.py [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+class CacheVolume(ExpObjVolume):
+    def __init__(self, cache, uuid=None):
+        super().__init__(cache, uuid)
+        self.cache = cache
+
+    # TODO: remove this mock after OCF properly handles
+    # cache device flush I/O with no data set
+    def do_submit_flush(self, flush):
+        io = cast(flush, POINTER(Io))
+        io.contents._end(io, 0)
```
Comment (robertbaldyga) [inline tests/functional/pyocf/types/volume_cache.py:23]: Patch already on review https://github.com/Open-CAS/ocf/pull/671

File: tests/functional/pyocf/types/volume_core.py [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+from pyocf.types.io import IoDir
+
+
+class CoreVolume(ExpObjVolume):
+    def __init__(self, core, uuid=None):
+        super().__init__(core, uuid)
+        self.core = core
+
+    def md5(self):
+        return self.core.exp_obj_md5()
```
Comment (robertbaldyga) [inline tests/functional/pyocf/types/volume_core.py:17]: Don't we want to move this functionality from `Core` to `CoreVolume`? `Core` is supposed to represent management object, while `CoreVolume` is I/O interface. Calculating MD5 sounds more like a feature of the latter.

**Review by robertbaldyga — CHANGES_REQUESTED** [review]

File: tests/functional/pyocf/types/core.py (hunk 18 above)
Comment (arutk) [reply tests/functional/pyocf/types/core.py:119]: done

File: tests/functional/pyocf/types/volume.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/pyocf/types/volume.py]: fixed?

File: tests/functional/pyocf/types/volume.py [hunk 21]
```diff
@@ -66,75 +66,132 @@ class VolumeProperties(Structure):
...
+                OcfLib.getInstance().ocf_io_get_volume(io_structure)
+            )
+
+            volume.submit_discard(io_structure)
+
+        @VolumeOps.SUBMIT_WRITE_ZEROES
+        def _submit_write_zeroes(write_zeroes):
+            raise NotImplementedError
+
+        @CFUNCTYPE(c_int, c_void_p)
```
Comment (jfckm) [inline tests/functional/pyocf/types/volume.py:125]: Why is this the only method, which is decorated generically and not with dedicated decorator like `VolumpeOps.OPEN`?

File: tests/functional/pyocf/types/volume_exp_obj.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jfckm) [reply tests/functional/pyocf/types/volume_exp_obj.py:39]: Oh, sorry, I didn't notice that there are two different IOs involved - I thought that it was just get data from IO and set it again with 0 offset.

File: tests/functional/pyocf/types/volume_core.py (hunk 20 above)
Comment (arutk) [reply tests/functional/pyocf/types/volume_core.py]: Ok, moved all I/O methods to volumes

File: tests/functional/pyocf/types/volume_cache.py (hunk 19 above)
Comment (arutk) [reply tests/functional/pyocf/types/volume_cache.py:27]: I will rebase in a separate push, to avoid mixing changes from baseline and dev branch in force-push preview

File: tests/functional/pyocf/types/core.py [hunk 22]
```diff
@@ -115,18 +115,23 @@ def new_io(
         return Io.from_pointer(io)
 
     def get_front_volume(self):
-        assert self.device == lib.ocf_core_get_volume(self.handle)
-        return lib.ocf_core_get_volume(self.handle)
+        return Volume.get_instance(lib.ocf_core_get_front_volume(self.handle))
+
+    def get_volume(self):
+        return Volume.get_instance(lib.ocf_core_get_volume(self.handle))
```
Comment (robertbaldyga) [inline tests/functional/pyocf/types/core.py:106]: This change does not seem to belong to this commit.

File: tests/functional/pyocf/types/volume_cache.py (hunk 19 above)
Comment (robertbaldyga) [reply tests/functional/pyocf/types/volume_cache.py:27]: This patch is already merged.

**Review by robertbaldyga — CHANGES_REQUESTED** [review]

File: tests/functional/pyocf/types/volume.py (hunk 21 above)
Comment (arutk) [reply tests/functional/pyocf/types/volume.py]: fixed

File: tests/functional/pyocf/types/volume_cache.py (hunk 19 above)
Comment (arutk) [reply tests/functional/pyocf/types/volume_cache.py]: ok, rebased and removed

**Review by robertbaldyga — APPROVED** [review]

**Review by jfckm — APPROVED** [review]
